libjulia/optimiser/FullInliner.cpp

```diff
@@ -0,0 +1,262 @@
+
+		m_statementsToPrefix.emplace_back(VariableDeclaration{
+			funCall.location,
+			{{funCall.location, variableReplacements[returnVariable], fun.returnVariables[0].type}},
+			{}
+		});
+		_expression = Identifier{funCall.location, variableReplacements[returnVariable]};
+	}
+	m_statementsToPrefix.emplace_back(BodyCopier(m_nameDispenser, fun.name + "_", variableReplacements)(fun.body));
+}
+
+void InlineModifier::visit(Statement& _statement)
+{
+	ASTModifier::visit(_statement);
+	// Replace pop(0) expression statemets (and others) by empty blocks.
+	if (_statement.type() == typeid(ExpressionStatement))
+	{
+		ExpressionStatement& expSt = boost::get<ExpressionStatement&>(_statement);
+		if (expSt.expression.type() == typeid(FunctionalInstruction))
+		{
+			FunctionalInstruction& funInstr = boost::get<FunctionalInstruction&>(expSt.expression);
+			if (funInstr.instruction == solidity::Instruction::POP)
+				if (MovableChecker(funInstr.arguments.at(0)).movable())
+					_statement = Block{expSt.location, {}};
+		}
+	}
+}
+
+void InlineModifier::visitArguments(
+	vector<Expression>& _arguments,
+	vector<string> const& _nameHints,
+	vector<string> const& _types,
+	bool _moveToFront
+)
+{
+	// If one of the elements moves parts to the front, all other elements right of it
+	// also have to be moved to the front to keep the order of evaluation.
+	vector<Statement> prefix;
+	for (size_t i = 0; i < _arguments.size(); ++i)
+	{
+		auto& arg = _arguments[i];
+		// TODO optimize vector operations, check that it actually moves
+		auto internalPrefix = visitRecursively(arg);
+		if (!internalPrefix.empty())
+		{
+			_moveToFront = true;
+			// We go through the arguments left to right, so we have to invert
+			// the prefixes.
+			prefix = std::move(internalPrefix) + std::move(prefix);
+		}
+		else if (_moveToFront)
+		{
+			auto location = locationOf(arg);
+			string var = newName(i < _nameHints.size() ? _nameHints[i] : "");
+			prefix.emplace(prefix.begin(), VariableDeclaration{
+				location,
+				{{TypedName{location, var, i < _types.size() ? _types[i] : ""}}},
+				make_shared<Expression>(std::move(arg))
+			});
+			arg = Identifier{location, var};
+		}
+	}
+	m_statementsToPrefix += std::move(prefix);
+}
+
+vector<Statement> InlineModifier::visitRecursively(Expression& _expression)
+{
+	vector<Statement> saved;
+	saved.swap(m_statementsToPrefix);
+	visit(_expression);
+	saved.swap(m_statementsToPrefix);
+	return saved;
+}
+
+string InlineModifier::newName(string const& _prefix)
+{
+	return m_nameDispenser.newName(_prefix);
+}
+
+Expression InlineModifier::noop(SourceLocation const& _location)
+{
+	return FunctionalInstruction{_location, solidity::Instruction::POP, {
+		Literal{_location, assembly::LiteralKind::Number, "0", ""}
+	}};
+}
+
+Statement BodyCopier::operator()(VariableDeclaration const& _varDecl)
+{
+	for (auto const& var: _varDecl.variables)
+		m_variableReplacements[var.name] = m_nameDispenser.newName(m_varNamePrefix + var.name);
```

Comment:
Either this should use the local wrapper `newName` or rather the wrapper should be removed.

Reply:
Sorry, sometimes it is confusing having all these different classes in the same file. Do we need the wrapper though?

Reply:
If you want, I can remove the wrapper also from the other class, I just found it more convenient.